mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp

```diff
@@ -0,0 +1,432 @@
+        }
+      }
+    }
+
+    // erase old operation
+    rewriter.eraseOp(op);
+
+    return success();
+  }
+};
+
+template <typename MQTGateDynOp, typename MQTGateOptOp>
+struct ConvertMQTDynGateOp : public OpConversionPattern<MQTGateDynOp> {
+  using OpConversionPattern<MQTGateDynOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(MQTGateDynOp op, typename MQTGateDynOp::Adaptor /*adaptor*/,
+                  ConversionPatternRewriter& rewriter) const override {
+    // get all the input qubits including the ctrl qubits
+    auto dynInQubitsValues = op.getInQubits();
+    auto dynPosCtrlQubitsValues = op.getPosCtrlInQubits();
+    auto dynNegCtrlQubitsValues = op.getNegCtrlInQubits();
+    auto dynAllQubits = op.getAllInQubits();
+    auto& qubitMap = getQubitMap();
+
+    // get the latest opt qubit of all dyn input qubits
+    std::vector<Value> optInQubits;
+    std::vector<Value> optPosCtrlQubitsValues;
+    std::vector<Value> optNegCtrlQubitsValues;
+
+    for (auto dynQubit : dynInQubitsValues) {
+      optInQubits.push_back(qubitMap[dynQubit]);
+    }
+    for (auto dynQubit : dynPosCtrlQubitsValues) {
+      optPosCtrlQubitsValues.push_back(qubitMap[dynQubit]);
+    }
+    for (auto dynQubit : dynNegCtrlQubitsValues) {
+      optNegCtrlQubitsValues.push_back(qubitMap[dynQubit]);
+    }
+
+    // append them to a single vector
+    std::vector<Value> allOptQubits(optInQubits.begin(), optInQubits.end());
+    allOptQubits.insert(allOptQubits.end(), optPosCtrlQubitsValues.begin(),
+                        optPosCtrlQubitsValues.end());
+    allOptQubits.insert(allOptQubits.end(), optNegCtrlQubitsValues.begin(),
+                        optNegCtrlQubitsValues.end());
+
+    // get the static params and paramMask if they exist
+    DenseF64ArrayAttr staticParams = nullptr;
+    if (auto optionalParams = op.getStaticParams()) {
+      staticParams = mlir::DenseF64ArrayAttr::get(rewriter.getContext(),
+                                                  optionalParams.value());
+    } else {
+      staticParams = DenseF64ArrayAttr{};
+    }
+    DenseBoolArrayAttr paramMask = nullptr;
+    if (auto optionalMask = op.getParamsMask()) {
+      paramMask = mlir::DenseBoolArrayAttr::get(rewriter.getContext(),
+                                                optionalMask.value());
+    } else {
+      paramMask = DenseBoolArrayAttr{};
+    }
+
+    // create new operation
+    Operation* mqtoptOp = rewriter.create<MQTGateOptOp>(
+        op.getLoc(), ValueRange(optInQubits).getTypes(),
+        ValueRange(optPosCtrlQubitsValues).getTypes(),
+        ValueRange(optNegCtrlQubitsValues).getTypes(), staticParams, paramMask,
+        op.getParams(), optInQubits, optPosCtrlQubitsValues,
+        optNegCtrlQubitsValues);
+
+    // iterate over all the dyn input qubits and update their latest opt qubit
+    Value dynQubit = nullptr;
+    for (size_t i = 0; i < dynAllQubits.size(); i++) {
+      dynQubit = dynAllQubits[i];
```

Comment:
```suggestion
    for (size_t i = 0; i < dynAllQubits.size(); i++) {
      const auto& dynQubit = dynAllQubits[i];
```
In my opinion, it makes the code slightly clearer and I think the compiler can optimize this a bit better.